File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/iam/dialog/action/SendDisplayedIamActionTest.java
package com.emarsys.mobileengage.iam.dialog.action;

import android.os.Handler;

import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.concurrency.CoreSdkHandlerProvider;
import com.emarsys.mobileengage.iam.InAppInternal;
import com.emarsys.testUtil.TimeoutUtils;
import com.emarsys.testUtil.mockito.ThreadSpy;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.Mockito;

import java.util.HashMap;
import java.util.Map;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class SendDisplayedIamActionTest {

    private static final String CAMPAIGN_ID = "123445";
    private static final String SID = "testSid";
    private static final String URL = "https://www.emarsys.com";

    private SendDisplayedIamAction action;

    private Handler handler;
    private InAppInternal inAppInternal;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void init() {
        handler = new CoreSdkHandlerProvider().provideHandler();
        inAppInternal = mock(InAppInternal.class);
        action = new SendDisplayedIamAction(handler, inAppInternal);
    }

    @After
    public void tearDown() {
        handler.getLooper().quit();
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_handler_mustNotBeNull() {
        new SendDisplayedIamAction(
                null,
                mock(InAppInternal.class)
        );
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_inAppInternal_mustNotBeNull() {
        new SendDisplayedIamAction(
                mock(Handler.class),
                null
        );
    }

    @Test(expected = IllegalArgumentException.class)
    public void testExecute_campaignIdMustNotBeNull() {
        action.execute(null, SID, URL);
    }

    @Test
    public void testExecute_callsRequestManager_withProperAttributes() {
        action.execute(CAMPAIGN_ID, SID, URL);

        String eventName = "inapp:viewed";
        Map<String, String> attributes = new HashMap<>();
        attributes.put("campaignId", CAMPAIGN_ID);
        attributes.put("sid", SID);
        attributes.put("url", URL);

        verify(inAppInternal, Mockito.timeout(500)).trackInternalCustomEventAsync(
                eventName,
                attributes,
                null);
    }

    @Test
    public void testExecute_callsRequestManager_withoutSidAndUrl_when_theyAreNull() throws Exception {
        action.execute(CAMPAIGN_ID, null, null);

        String eventName = "inapp:viewed";
        Map<String, String> attributes = new HashMap<>();
        attributes.put("campaignId", CAMPAIGN_ID);

        verify(inAppInternal, Mockito.timeout(500)).trackInternalCustomEventAsync(eventName, attributes, null);
    }

    @Test
    @SuppressWarnings("unchecked")
    public void testExecute_callsRequestManager_onCoreSdkThread() {
        ThreadSpy threadSpy = new ThreadSpy();
        doAnswer(threadSpy).when(inAppInternal).trackInternalCustomEventAsync(
                any(String.class),
                any(Map.class),
                (CompletionListener) isNull());

        action.execute(CAMPAIGN_ID, SID, URL);

        threadSpy.verifyCalledOnCoreSdkThread();
    }


}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/iam/dialog/action/SendDisplayedIamActionTest.kt
package com.emarsys.mobileengage.iam.dialog.action


import com.emarsys.core.concurrency.ConcurrentHandlerHolderFactory.create
import com.emarsys.core.handler.ConcurrentHandlerHolder
import com.emarsys.mobileengage.iam.InAppInternal
import com.emarsys.testUtil.mockito.ThreadSpy
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.doAnswer
import org.mockito.kotlin.isNull
import org.mockito.kotlin.whenever

class SendDisplayedIamActionTest  {
    private lateinit var action: SendDisplayedIamAction
    private lateinit var handler: ConcurrentHandlerHolder
    private lateinit var inAppInternal: InAppInternal

    @Before
    fun init() {
        handler = create()
        inAppInternal = Mockito.mock(InAppInternal::class.java)
        action = SendDisplayedIamAction(handler, inAppInternal)
    }

    @After
    fun tearDown() {
        handler.coreLooper.quit()
    }

    @Test
    fun testExecute_callsRequestManager_withProperAttributes() {
        action!!.execute(CAMPAIGN_ID, SID, URL)
        val eventName = "inapp:viewed"
        val attributes: MutableMap<String, String> = HashMap()
        attributes["campaignId"] = CAMPAIGN_ID
        attributes["sid"] = SID
        attributes["url"] = URL
        Mockito.verify(inAppInternal, Mockito.timeout(500)).trackInternalCustomEventAsync(
            eventName,
            attributes,
            null
        )
    }

    @Test
    @Throws(Exception::class)
    fun testExecute_callsRequestManager_withoutSidAndUrl_when_theyAreNull() {
        action.execute(CAMPAIGN_ID, null, null)
        val eventName = "inapp:viewed"
        val attributes: MutableMap<String, String> = HashMap()
        attributes["campaignId"] = CAMPAIGN_ID
        Mockito.verify(inAppInternal, Mockito.timeout(500))
            .trackInternalCustomEventAsync(eventName, attributes, null)
    }

    @Test
    fun testExecute_callsRequestManager_onCoreSdkThread() {
        val threadSpy: ThreadSpy<Any> = ThreadSpy()
        whenever(inAppInternal.trackInternalCustomEventAsync(any(), any(), isNull())) doAnswer {
            threadSpy.call()
            null
        }

        action.execute(CAMPAIGN_ID, SID, URL)
        threadSpy.verifyCalledOnCoreSdkThread()
    }

    companion object {
        private const val CAMPAIGN_ID = "123445"
        private const val SID = "testSid"
        private const val URL = "https://www.emarsys.com"
    }
}